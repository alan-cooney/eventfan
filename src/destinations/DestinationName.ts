/* eslint-disable import/prefer-default-export */

/**
 * Destination names
 */
export enum DestinationName {
  FACEBOOK_PIXEL = "FACEBOOK_PIXEL",
  HOTJAR = "HOTJAR",
  GA4 = "GA4",
  POSTHOG = "POSTHOG",
}
