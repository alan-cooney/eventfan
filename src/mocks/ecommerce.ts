import {
  CartShared,
  CartViewed,
  CheckoutStarted,
  CheckoutStepCompleted,
  CheckoutStepViewed,
  OrderCompleted,
  OrderRefunded,
  PaymentInfoEntered,
  ProductAdded,
  ProductClicked,
  ProductListViewed,
  ProductRemoved,
  ProductShared,
  ProductsSearched,
  ProductViewed,
  PromotionClicked,
  PromotionViewed,
} from "../sourceEvents/ecommerce";

/**
 * Helper Mock Product
 */
const mockProduct: ProductViewed["properties"] = {
  category: "Trip",
  name: "Product Name",
  price: 100.0,
  product_id: "productID",
  quantity: 1,
  sku: "sku",
  variant: "variant",
};

export const mockCartShared: CartShared = {
  name: "Cart Shared",
  properties: {
    cart_id: "cartID",
    products: [mockProduct],
    recipient: "recipient",
    share_message: "share message",
    share_via: "Facebook",
  },
};

export const mockCartViewed: CartViewed = {
  name: "Cart Viewed",
  properties: {
    cart_id: "cartID",
    products: [mockProduct],
  },
};

export const mockCheckoutStarted: CheckoutStarted = {
  name: "Checkout Started",
  properties: {
    currency: "GBP",
    order_id: "orderID",
    products: [mockProduct],
    revenue: 100.0,
    total: 200.0,
  },
};

export const mockCheckoutStepCompleted: CheckoutStepCompleted = {
  name: "Checkout Step Completed",
  properties: {
    checkout_id: "checkoutID",
    payment_method: "Credit Card",
    shipping_method: "shippingMethod",
    step: 1,
  },
};

export const mockCheckoutStepViewed: CheckoutStepViewed = {
  name: "Checkout Step Viewed",
  properties: mockCheckoutStepCompleted.properties,
};

export const mockOrderCompleted: OrderCompleted = {
  name: "Order Completed",
  properties: mockCheckoutStarted.properties,
};

export const mockOrderRefunded: OrderRefunded = {
  name: "Order Refunded",
  properties: mockCheckoutStarted.properties,
};

export const mockPaymentInfoEntered: PaymentInfoEntered = {
  name: "Payment Info Entered",
  properties: {
    currency: "GBP",
    order_id: "orderID",
    products: [mockProduct],
    revenue: 100.0,
  },
};

export const mockProductAdded: ProductAdded = {
  name: "Product Added",
  properties: mockProduct,
};

export const mockProductClicked: ProductClicked = {
  name: "Product Clicked",
  properties: mockProduct,
};

export const mockProductListViewed: ProductListViewed = {
  name: "Product List Viewed",
  properties: {
    category: "category",
    list_id: "list_id",
    products: [mockProduct],
  },
};

export const mockProductRemoved: ProductRemoved = {
  name: "Product Removed",
  properties: mockProduct,
};

export const mockProductsShared: ProductShared = {
  name: "Product Shared",
  properties: {
    ...mockProduct,
    recipient: "recipient",
    share_message: "share message",
    share_via: "Facebook",
  },
};

export const mockProductsSearched: ProductsSearched = {
  name: "Products Searched",
  properties: {
    query: "Query string",
  },
};

export const mockProductViewed: ProductViewed = {
  name: "Product Viewed",
  properties: mockProduct,
};

export const mockPromotionClicked: PromotionClicked = {
  name: "Promotion Clicked",
  properties: {
    promotion_creative: "creative",
    promotion_id: "promotionID",
    promotion_name: "name",
    promotion_position: "position",
  },
};

export const mockPromotionViewed: PromotionViewed = {
  name: "Promotion Viewed",
  properties: mockPromotionClicked.properties,
};
