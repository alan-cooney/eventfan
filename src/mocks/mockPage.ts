import { Page } from "../types/PageViewProps";

const mockPage: Page = {
  options: undefined,
  name: "title",
  properties: {
    path: "/path",
    title: "title",
    url: "https://www.example.com/path",
  },
};

export default mockPage;
