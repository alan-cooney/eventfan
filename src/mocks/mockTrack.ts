import { TEvent } from "../types/TrackEvent";

const mockTrack: TEvent = {
  name: "Test Event",
  properties: {
    iceCream: "vanilla",
  },
};

export default mockTrack;
