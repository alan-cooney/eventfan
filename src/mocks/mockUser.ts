import { User } from "../types/User";

const mockUser: User = {
  options: undefined,
  traits: {
    email: "test@gmail.com",
    firstName: "First Name",
  },
  userId: "userID",
};

export default mockUser;
