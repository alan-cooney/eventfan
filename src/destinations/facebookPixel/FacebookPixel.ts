/* eslint-disable @typescript-eslint/no-explicit-any */
import { IdentifyProps } from "../../types/IdentifyProps";
import { TEvent } from "../../types/TrackEvent";
import Destination from "../Destination";
import { DestinationName } from "../DestinationName";
import loadScript from "../loadScript";
import orderCompleted from "./mapping/ecommerce/orderCompleted";
import { AdvancedMatching } from "./types/AdvancedMatching";

export default class FacebookPixel implements Destination {
  private fb = (window as any).fbq;

  constructor(private pixelID: string) {}

  eventMappings = {
    "Order Completed": orderCompleted,
  };

  identify(user: IdentifyProps): void {
    const traits: AdvancedMatching = {
      country: user.address?.country?.toLowerCase(),
      fn: user.firstName,
      ln: user.lastName,
      ct: user.address?.city,
      //   db: user.birthday,
      external_id: user.id,
      //   ge: user.gender,
      //   ph: parseInt(user.phone?),
      st: user.address?.state,
      zp: user.address?.state,
    };

    // Re-initialise with advanced matching parameters
    this.fb("init", "PageView", traits);
  }

  async initialise(): Promise<void> {
    await loadScript(
      "facebook-pixel-integration",
      "https://connect.facebook.net/en_US/fbevents.js"
    );
    this.fb("init", this.pixelID);
  }

  page(): void {
    this.fb("track", "PageView");
  }

  track(event: TEvent): Promise<void> {
    this.fb("track", event.eventName, event.props);

    // Resolve immediately as Facebook Pixel has no way of firing a callback once complete
    return Promise.resolve();
  }

  name = DestinationName.FACEBOOK_PIXEL;

  isLoaded: boolean = typeof window !== "undefined" && !!(window as any).fbq;
}
