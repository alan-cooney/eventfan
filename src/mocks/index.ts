import mockPage from "./mockPage";
import mockTrack from "./mockTrack";
import mockUser from "./mockUser";

export { mockPage, mockTrack, mockUser };
export * from "./ecommerce";
