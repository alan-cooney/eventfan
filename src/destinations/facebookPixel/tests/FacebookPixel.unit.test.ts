/* eslint-disable @typescript-eslint/dot-notation */
import { mockTrack, mockUser } from "../../../mocks";
import { DestinationName } from "../../DestinationName";
import FacebookPixel from "../FacebookPixel";

it("has the correct destination name", () => {
  const fb = new FacebookPixel({ pixelId: "pixelId" });
  expect(fb.name).toBe(DestinationName.FACEBOOK_PIXEL);
});

it("starts as not loaded", () => {
  const fb = new FacebookPixel({ pixelId: "pixelId" });
  expect(fb.isLoaded).toBeFalsy();
});

describe("identify", () => {
  it("formats the advance matching parameters in a way that matches the snapshot", async () => {
    const fb = new FacebookPixel({ pixelId: "pixelId" });
    const mockFbWindowObject = jest.fn() as any;
    fb["fb"] = mockFbWindowObject;

    await fb.identify(mockUser);

    expect(mockFbWindowObject.mock.calls[0]).toMatchSnapshot();
  });
});

describe("page", () => {
  it("calls the correct facebook track method", async () => {
    const fb = new FacebookPixel({ pixelId: "pixelId" });
    const mockFbWindowObject = jest.fn() as any;
    fb["fb"] = mockFbWindowObject;

    await fb.page();

    expect(mockFbWindowObject.mock.calls[0]).toMatchSnapshot();
  });
});

describe("track", () => {
  it("calls the facebook track method with parameters matching the snapshot", async () => {
    const fb = new FacebookPixel({ pixelId: "pixelId" });
    const mockFbWindowObject = jest.fn() as any;
    fb["fb"] = mockFbWindowObject;

    await fb.track(mockTrack);

    expect(mockFbWindowObject.mock.calls[0]).toMatchSnapshot();
  });
});
