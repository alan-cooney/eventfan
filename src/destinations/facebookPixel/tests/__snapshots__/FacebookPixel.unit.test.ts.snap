// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`identify formats the advance matching parameters in a way that matches the snapshot 1`] = `
[
  "init",
  "pixelId",
  {
    "country": "gb",
    "ct": "city",
    "db": 19900101,
    "em": "test@gmail.com",
    "external_id": "userID",
    "fn": "firstname",
    "ge": "m",
    "ln": "lastname",
    "ph": 447111111111,
    "st": "london",
    "zp": "SW11 111",
  },
]
`;

exports[`page calls the correct facebook track method 1`] = `
[
  "track",
  "PageView",
]
`;

exports[`track calls the facebook track method with parameters matching the snapshot 1`] = `
[
  "track",
  "Purchase",
  {
    "currency": "GBP",
    "order_id": "orderID",
    "products": [
      {
        "category": "Trip",
        "currency": "GBP",
        "image_url": "http://example.com/image.jpg",
        "name": "Product Name",
        "price": 100,
        "product_id": "productID",
        "quantity": 1,
        "sku": "sku",
        "url": "http://example.com/productID",
        "variant": "variant",
      },
    ],
    "revenue": 100,
    "total": 200,
  },
]
`;

exports[`track calls the facebook trackCustom method with parameters matching the snapshot 1`] = `
[
  "trackCustom",
  "Test Event",
  {
    "iceCream": "vanilla",
  },
]
`;
